File: BigWorldServer/world/PlayerConfigTable.go
package main

import (
	"os"

	pb "bigworld/common/pb"
)

type PlayerConfigTable struct {
	maxStep           float64
	sprintToRunTime   float64
	fallGravityMps2   float64
	fallSpeedLimitMps float64
	playerHeight      float64
	playerCenterY     float64
}

var playerConfigTable = &PlayerConfigTable{
	maxStep:           0.5,
	sprintToRunTime:   1.0,
	fallGravityMps2:   10.0,
	fallSpeedLimitMps: 15.0,
	playerHeight:      1.8,
	playerCenterY:     0.9,
}

func PlayerConfig() *PlayerConfigTable { return playerConfigTable }

func (c *PlayerConfigTable) Load(path string) error {
	data, err := os.ReadFile(path)
	if err != nil {
		return err
	}
	msg := pb.GetRootAsPlayerConfigMsg(data, 0)
	c.maxStep = float64(msg.VoxelMaxStepHeight())
	c.sprintToRunTime = float64(msg.SprintToRunTime())
	c.fallGravityMps2 = float64(msg.Gravity())
	c.fallSpeedLimitMps = float64(msg.FallSpeedLimit())
	c.playerHeight = float64(msg.ColliderHeight())
	c.playerCenterY = float64(msg.ColliderCenterY())

	if c.maxStep <= 0 {
		c.maxStep = 0.5
	}
	if c.sprintToRunTime <= 0 {
		c.sprintToRunTime = 1.0
	}
	if c.fallGravityMps2 <= 0 {
		c.fallGravityMps2 = 10.0
	}
	if c.fallSpeedLimitMps <= 0 {
		c.fallSpeedLimitMps = 15.0
	}
	if c.playerHeight <= 0 {
		c.playerHeight = 1.8
	}
	if c.playerCenterY <= 0 {
		c.playerCenterY = 0.9
	}
	return nil
}

func (c *PlayerConfigTable) MaxStep() float64           { return c.maxStep }
func (c *PlayerConfigTable) SprintToRunTime() float64   { return c.sprintToRunTime }
func (c *PlayerConfigTable) FallGravityMps2() float64   { return c.fallGravityMps2 }
func (c *PlayerConfigTable) FallSpeedLimitMps() float64 { return c.fallSpeedLimitMps }
func (c *PlayerConfigTable) PlayerHeight() float64      { return c.playerHeight }
func (c *PlayerConfigTable) PlayerCenterY() float64     { return c.playerCenterY }

File: BigWorldClient/Assets/Config/PlayerConfigTable.cs
using System.Collections.Generic;
using BigWorldClient.Network.Protocol;
using Google.FlatBuffers;
using UnityEngine;

namespace BigWorldClient
{

    public sealed class PlayerConfigTable
    {
        private static PlayerConfigTable _cached;

        public float SprintToRunTime { get; private set; } = 1f;
        public float FallSpeedLimit { get; private set; } = 15f;
        public float Gravity { get; private set; } = 1f;
        public float ColliderHeight { get; private set; } = 1.8f;
        public float ColliderCenterY { get; private set; } = 0.9f;
        public float ColliderRadius { get; private set; } = 0.2f;
        public float StepHeightPercentage { get; private set; } = 0.25f;
        public float VoxelMaxStepHeight { get; private set; } = 0.5f;

        public List<CameraRecenteringEntry> BackwardsRecenteringData { get; private set; } = new();
        public List<CameraRecenteringEntry> SidewaysRecenteringData { get; private set; } = new();

        public static PlayerConfigTable Instance => _cached ??= Load();

        static PlayerConfigTable Load()
        {
            var table = new PlayerConfigTable();
            TextAsset asset = Resources.Load<TextAsset>("Config/PlayerConfig");
            if (asset != null)
            {
                PlayerConfigMsg msg = PlayerConfigMsg.GetRootAsPlayerConfigMsg(new ByteBuffer(asset.bytes));
                table.SprintToRunTime = msg.SprintToRunTime;
                table.FallSpeedLimit = msg.FallSpeedLimit;
                table.Gravity = msg.Gravity;
                table.ColliderHeight = msg.ColliderHeight;
                table.ColliderCenterY = msg.ColliderCenterY;
                table.ColliderRadius = msg.ColliderRadius;
                table.StepHeightPercentage = msg.StepHeightPercentage;
                table.VoxelMaxStepHeight = msg.VoxelMaxStepHeight;
                var backwards = new List<CameraRecenteringEntry>(msg.CameraBackwardsLength);
                for (int i = 0; i < msg.CameraBackwardsLength; i++)
                {
                    var e = msg.CameraBackwards(i);
                    if (e.HasValue) backwards.Add(e.Value);
                }
                table.BackwardsRecenteringData = backwards;
                var sideways = new List<CameraRecenteringEntry>(msg.CameraSidewaysLength);
                for (int i = 0; i < msg.CameraSidewaysLength; i++)
                {
                    var e = msg.CameraSideways(i);
                    if (e.HasValue) sideways.Add(e.Value);
                }
                table.SidewaysRecenteringData = sideways;
            }

            return table;
        }
    }
}
